File: chapter1/c++/4.cpp
/*
Given a string, write a function to check if it is a permutation of a palindrome. A palindrome is a word or phrase that is the same forwards and backwards. A permutation is a rearrangement of letters. The palindrome does not need to be limited to just dictionary words. You can ignore casing and non-letter characters.

	EXAMPLE

	Input: Tact Coa
	Output: True (permutations: "taco cat", "atco cta", etc.)
*/

#include <bits/stdc++.h>
using namespace std;

bool isPermutationOfPalindrome(string str) {
	string strNoSpace;
	for(int i=0; i<str.size(); i++) {
		if(str[i]!=' ') strNoSpace += (char)tolower(str[i]);
	}

	map<char,int> m;
	for(int i=0; i<strNoSpace.size(); i++) {
		if(m.find(strNoSpace[i])==m.end()) {
			m.insert(make_pair(strNoSpace[i], 1));
		} else {
			m[strNoSpace[i]]++;
		}
	}
	
	bool odd = strNoSpace.size()%2!=0;
	bool encountered_solo = false;
	map<char,int>::iterator itr;
	for(itr = m.begin(); itr != m.end(); itr++) {
		if(itr->second%2!=0) {
			if(odd) {
				if(encountered_solo) return false;
				else encountered_solo = true;
			} else
				return false;
		}
	}
	return true;
}

int main() {
	cout << isPermutationOfPalindrome("Tact Coan");
}

File: chapter1/4.py
'''
Given a string, write a function to check if it is a permutation of a palindrome.
A palindrome is a word or phrase that is the same forwards and backwards.
A permutation is a rearrangement of letters. The palindrome does not need
to be limited to just dictionary words. You can ignore casing and non-letter characters.

	EXAMPLE

	Input: Tact Coa
	Output: True (permutations: "taco cat", "atco cta", etc.)

'''
def is_permutation_of_palindrome(str: str):
	str = str.lower()
	c_set = dict()

	for c in str:
		if c in c_set:
			c_set[c] += 1
		else:
			c_set.update({c: 1})

	odd = len(str)%2==0
	encountered_solo = False
	#del c_set[' ']
	for key in c_set:
		if key == ' ':
			continue
		if c_set[key]%2 != 0:
			if odd:
				if encountered_solo:
					return False
				else:
					encountered_solo = True
			else:
				return False
	return True


def main():
	print(is_permutation_of_palindrome("Tact Coa"))

if __name__ == "__main__":
	main()